File: clients/ios/Lextures/Core/LMS/ModuleContentLogic.swift
import Foundation

// MARK: - Module grouping

struct ModuleGroup: Identifiable, Equatable {
    let id: String
    let title: String
    let items: [CourseStructureItem]
}

enum ModuleItemDestination: Equatable {
    case contentPage
    case quiz
    case assignment
    case externalLink
    case webContent
    case interactive
    case file
    case unsupported
}

enum ModuleContentLogic {
    static func buildModuleGroups(from items: [CourseStructureItem]) -> [ModuleGroup] {
        let modules = items.filter(\.isModule).sorted { $0.sortOrder < $1.sortOrder }
        let children = Dictionary(grouping: items.filter { !$0.isModule && $0.parentId != nil }) { $0.parentId! }
        var groups = modules.map { module in
            ModuleGroup(
                id: module.id,
                title: module.title,
                items: (children[module.id] ?? []).sorted { $0.sortOrder < $1.sortOrder }
            )
        }
        let orphans = items
            .filter { !$0.isModule && $0.parentId == nil && $0.kind != "heading" }
            .sorted { $0.sortOrder < $1.sortOrder }
        if !orphans.isEmpty {
            groups.append(ModuleGroup(id: "__orphans__", title: "Other items", items: orphans))
        }
        return groups
    }

    static func destination(for kind: String) -> ModuleItemDestination {
        switch kind {
        case "content_page": return .contentPage
        case "quiz": return .quiz
        case "assignment": return .assignment
        case "external_link": return .externalLink
        case "textbook_resource", "library_resource": return .webContent
        case "h5p", "scorm", "lti_link", "vibe_activity": return .interactive
        case "file", "file_item": return .file
        default: return .unsupported
        }
    }

    static func isNavigable(_ kind: String) -> Bool {
        switch destination(for: kind) {
        case .unsupported: return false
        default: return true
        }
    }

    static func itemLockState(in progress: ModulesProgressSnapshot?, itemId: String) -> ItemLockState? {
        guard let progress else { return nil }
        for module in progress.modules {
            for item in module.items ?? [] where item.itemId == itemId {
                return item
            }
        }
        return nil
    }

    static func moduleLockState(in progress: ModulesProgressSnapshot?, moduleId: String) -> ModuleLockState? {
        progress?.modules.first { $0.moduleId == moduleId }
    }

    static func isLocked(in progress: ModulesProgressSnapshot?, itemId: String) -> Bool {
        itemLockState(in: progress, itemId: itemId)?.locked == true
    }

    static func isComplete(in progress: ModulesProgressSnapshot?, itemId: String) -> Bool {
        itemLockState(in: progress, itemId: itemId)?.complete == true
    }

    static func accessibilityLabel(for item: CourseStructureItem, progress: ModulesProgressSnapshot?) -> String {
        var parts = [ItemKind.label(for: item.kind), item.title]
        if isComplete(in: progress, itemId: item.id) {
            parts.append(L.text("mobile.modules.complete"))
        }
        if isLocked(in: progress, itemId: item.id),
           let reason = itemLockState(in: progress, itemId: item.id)?.reason?.message,
           !reason.isEmpty {
            parts.append(reason)
        }
        return parts.joined(separator: ", ")
    }
}

File: clients/android/app/src/main/kotlin/com/lextures/android/core/lms/ModuleContentLogic.kt
package com.lextures.android.core.lms

data class ModuleGroup(
    val id: String,
    val title: String,
    val items: List<CourseStructureItem>,
)

enum class ModuleItemDestination {
    ContentPage,
    Quiz,
    Assignment,
    ExternalLink,
    WebContent,
    Interactive,
    VibeActivity,
    LibraryResource,
    File,
    Unsupported,
}

object ModuleContentLogic {
    fun buildModuleGroups(items: List<CourseStructureItem>): List<ModuleGroup> {
        val modules = items.filter { it.isModule }.sortedBy { it.sortOrder }
        val children = items.filter { !it.isModule && it.parentId != null }.groupBy { it.parentId }
        val grouped = modules.map { module ->
            ModuleGroup(module.id, module.title, (children[module.id] ?: emptyList()).sortedBy { it.sortOrder })
        }
        val orphans = items
            .filter { !it.isModule && it.parentId == null && it.kind != "heading" }
            .sortedBy { it.sortOrder }
        return if (orphans.isEmpty()) grouped else grouped + ModuleGroup("__orphans__", "Other items", orphans)
    }

    fun destination(kind: String): ModuleItemDestination = when (kind) {
        "content_page" -> ModuleItemDestination.ContentPage
        "quiz" -> ModuleItemDestination.Quiz
        "assignment" -> ModuleItemDestination.Assignment
        "external_link" -> ModuleItemDestination.ExternalLink
        "library_resource" -> ModuleItemDestination.LibraryResource
        "textbook_resource" -> ModuleItemDestination.WebContent
        "h5p", "scorm", "lti_link" -> ModuleItemDestination.Interactive
        "vibe_activity" -> ModuleItemDestination.VibeActivity
        "file", "file_item" -> ModuleItemDestination.File
        else -> ModuleItemDestination.Unsupported
    }

    fun isNavigable(kind: String): Boolean = destination(kind) != ModuleItemDestination.Unsupported

    fun itemLockState(progress: ModulesProgressSnapshot?, itemId: String): ItemLockState? {
        progress?.modules.orEmpty().forEach { module ->
            module.items.orEmpty().firstOrNull { it.itemId == itemId }?.let { return it }
        }
        return null
    }

    fun moduleLockState(progress: ModulesProgressSnapshot?, moduleId: String): ModuleLockState? =
        progress?.modules.orEmpty().firstOrNull { it.moduleId == moduleId }

    fun isLocked(progress: ModulesProgressSnapshot?, itemId: String): Boolean =
        itemLockState(progress, itemId)?.locked == true

    fun isComplete(progress: ModulesProgressSnapshot?, itemId: String): Boolean =
        itemLockState(progress, itemId)?.complete == true
}
